warehouse/oidc/forms/gitlab.py

```diff
@@ -0,0 +1,111 @@
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+import wtforms
+
+from warehouse import forms
+from warehouse.i18n import localize as _
+from warehouse.oidc.forms._core import PendingPublisherMixin
+
+# TODO Gitlab also accepts emojis
+# see:
+# https://docs.gitlab.com/ee/user/reserved_names.html#limitations-on-project-and-group-names
+_VALID_GITLAB_PROJECT = re.compile(r"^[a-zA-Z0-9_][a-zA-Z0-9-_.+\- ]*$")
+_VALID_GITLAB_NAMESPACE = re.compile(r"^[a-zA-Z0-9_][a-zA-Z0-9_.()\- /]*$")
+_VALID_GITLAB_ENVIRONMENT = re.compile(r"^[a-zA-Z0-9\-_/${} ]+$")
+
+
+class GitLabPublisherBase(forms.Form):
+    __params__ = ["namespace", "project", "workflow_filepath", "environment"]
+
+    namespace = wtforms.StringField(
+        validators=[
+            wtforms.validators.InputRequired(
+                message=_("Specify GitLab namespace (username or group/subgroup)"),
+            ),
+        ]
+    )
+
+    project = wtforms.StringField(
+        validators=[
+            wtforms.validators.InputRequired(message=_("Specify project name")),
+            wtforms.validators.Regexp(
+                _VALID_GITLAB_PROJECT, message=_("Invalid project name")
+            ),
+        ]
+    )
+
+    workflow_filepath = wtforms.StringField(
+        validators=[
+            wtforms.validators.InputRequired(message=_("Specify workflow filepath"))
+        ]
+    )
+
+    environment = wtforms.StringField(
+        validators=[
+            wtforms.validators.Optional(),
+            wtforms.validators.Regexp(
+                _VALID_GITLAB_ENVIRONMENT, message=_("Invalid environment name")
+            ),
+        ]
+    )
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def validate_namespace(self, field):
+        namespace = field.data
+
+        # We pre-filter namespace with a regex
+        if not _VALID_GITLAB_NAMESPACE.match(namespace):
+            raise wtforms.validators.ValidationError(
+                _("Invalid GitLab username or group/subgroup name.")
+            )
```

Comment:
Can this be an inline `validators=[]` member instead? Might be missing something.

Reply:
yup, you're right, the separate function was a leftover from how GitHub does it. Fixed.